rfcs/0002-borrow-type.md

```diff
@@ -0,0 +1,277 @@
+
+# Guide-level Explanation  
+[guide-level-explanation]: #guide-level-explanation
+
+***Note**: The following is what documentation for this feature might look like.*
+
+You may see the type `Borrow<'a, T>` used in Windows APIs. There's no need to worry about this type, it's actually quite straight forward.
+
+`Borrow<'a, T>` is exactly equivalent to `&'a T` except that in memory a `Borrow<'a, T>` looks just like a `T`.
+
+This is important because some Windows APIs take arguments of some type `T` but treat them as they are just borrowed.
+
+Let's take a look at an example:
+
+`windows::Data::Json::JsonObject` has a function named `Parse` which looks like this:
+
+```rust 
+fn Parse<'a>(input: Borrow<'a, HSTRING>) -> Result<JsonObject>;
+```
+
+The reason `Parse` takes a `Borrow<'a, HSTRING>` instead of just a plain `HSTRING` is because if it did, ownership of `input` would be passed into `Parse` and `input`'s `Drop` implementation would be called possibly freeing `input`'s memory!
+
+`Parse` really expects to logically borrow `input`. If this were a normal Rust function, you would expect `input` to be something like `&HSTRING`. However, `Parse` is not a regular Rust function - it crosses the FFI boundary into another language entirely! If were were to try to pass a `&HSTRING` into `Parse`, `Parse` would interpret the reference as if it were just a plain old `HSTRING`. This would cause all kinds of problem's all caused by the fact that `Parse` expected an `HSTRING` and we tried to give it a `&HSTRING`.
+
+`Borrow<'a, T>` allows us to emulate a reference but at the memory level still pass around memory that looks just like the underlying `T`. We get the safety of a Rust reference when writing our Rust code, and the Windows API receives the memory it expects to. Everyone wins!
+
+# Reference-level Explanation  
+[reference-level-explanation]: #reference-level-explanation
+
+`Borrow<'a, T>` has the following definition:
+
+```rust
+#![warn(unsafe_op_in_unsafe_fn)]
+
+#[repr(transparent)]
+struct Borrow<'a, T> {
+    item: core::mem::ManuallyDrop<T>,
+    lifetime: core::marker::PhantomData<&'a ()>
+}
+
+impl <'a, T> Borrow<'a, T> {
+    /// # Safety
+    ///
+    /// It must be safe to alias `T` as long as `T`'s 
+    /// destructor is not run. 
+    ///
+    /// For example, `IUnknown` is fine to alias as long as 
+    /// the reference count is guaranteed to stay above 0 
+    /// until the original `item` is dropped. On the other hand
+    /// `Box<T>` is not fine to alias and thus passing a `Box` 
+    /// to this function would be UB.
+    unsafe fn new(item: &'a T) -> Self {
+        let item = unsafe { core::mem::transmute_copy(item) };
+        Self {
+            item,
+            lifetime: core::marker::PhantomData
+        }
+    }
+}
+
+impl<'a, T> std::ops::Deref for Borrow<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.item
+    }
+}
+```
+
+## Usages
```

Comment:
I will add this to the document, but `Option<Borrowed<'a, IUnknown>>` would have the same layout as `Option<IUnknown>` (i.e., a pointer), so that should not be a problem.

Reply:
I was just looking for something a little more fluent, perhaps you can construct a Borrowed from a None and it would produce a zeroed Borrowed value. The thing is that practically everywhere you'd use a Borrowed you would also be expected to accept a zeroed/none value.